proposals/subgroups.md

```diff
@@ -0,0 +1,227 @@
+| `fn subgroupMin(e : T) -> T` | `T` must be u32, i32, f32, f16 or a vector of those types | Reduction<br>Performs a min of `e` among all active invocations and returns that result |
+| `fn subgroupMax(e : T) -> T` | `T` must be u32, i32, f32, f16 or a vector of those types | Reduction<br>Performs a max of `e` among all active invocations and returns that result |
+| `fn quadBroadcast(e : T, id : I)` | `T` must be u32, i32, f32, f16 or a vector of those types<br>`I` must be u32 or i32 | Broadcasts `e` from the quad invocation with id equal to `id`<br>`e` must be a constant-expression<sup>2</sup> |
+| `fn quadSwapX(e : T)` | `T` must be u32, i32, f32, f16 or a vector of those types | Swaps `e` between invocations in the quad in the X direction |
+| `fn quadSwapY(e : T)` | `T` must be u32, i32, f32, f16 or a vector of those types | Swaps `e` between invocations in the quad in the Y direction |
+| `fn quadSwapDiagonal(e : T)` | `T` must be u32, i32, f32, f16 or a vector of those types | Swaps `e` between invocations in the quad diagnoally |
+1. This is the first instance of dynamic uniformity. See the portability and uniformity section for more details.
+2. Unlike `subgroupBroadcast`, SPIR-V does not have a shuffle operation to fall back on, so this requirement must be surfaced.
+
+**TODO**: Are quad operations worth it?
+Quad operations present even less portability than subgroup operations due to
+factors like helper invocations and multiple draws being packed into a
+subgroup.
+SM6.7 adds an attribute to require helpers be active.
+
+## Portability and Uniformity
+
+Unfortunately,
+[testing](https://github.com/gpuweb/gpuweb/issues/4306#issuecomment-1795498468)
+indicates that behavior is not widely portable across devices.
+Even requiring that the subgroup operations only be used in uniform control
+flow (at workgroup scope) is insufficient to produce portable behavior.
+For example, compilers make aggressive opimizations that do not preserve the
+correct active invocations.
+This leaves us in an awkward situation where portability cannot be guaranteed,
+but these operations provide significant performance improvements in many
+circumstances.
+
+Suggest allowing these operations anywhere and provide guidance on how to
+achieve portable behavior.
+From testing, it seems all implementations are able to produce portable results
+when the workgroup never diverges.
+While this may seem obvious, it still provides significant benefit in many
+cases (for example, by reducing overall memory bandwidth).
+
+Not requiring any particular uniformity also makes providing these operations
+in fragment shaders more palatable.
+Normally, there would be extra portability hazards in fragment shaders (e.g.
+due to helper invocations).
+
+**TODO**: Is it worth adding a diagnostic control to require subgroup
```

Comment:
It would be really nice. Even if we are building on shaky foundations semantics-wise, I think it is useful to help developers that want to build on stronger format foundations. It might be worth defining subgroup uniformity as well so we can require that bradcast's I is subgroup uniform, even if for now the only way to have that is by having I be workgroup uniform.

Reply:
As I replied to Greg's comment we could also consider a diagnostic that warns if all the branches are uniform (as opposed to just the subgroup operation being in uniform control flow).

I think adding either or both diagnostics defaulting to off is reasonable

Reply:
I've added a section on diagnostics that would default to errors as a starting point.